x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/create_upgradeable_rules_payload.ts

```diff
@@ -0,0 +1,190 @@
+
+      let rulePayload: Partial<PrebuiltRuleAsset> = {};
+
+      for (const field of fieldsPayload) {
+        rulePayload = {
+          ...rulePayload,
+          [field]: getValueForField({
+            field,
+            upgradeableRule,
+            globalPickVersion,
+            requestBody,
+            ruleId,
+            ruleFieldsDiff: calculatedRuleDiff,
+          }),
+        };
+      }
+
+      ruleUpgradePayloads.push(rulePayload as PrebuiltRuleAsset);
+    } catch (err) {
+      payloadErrors.push({
+        error: err,
+        item: { rule_id: ruleId },
+      });
+    }
+  });
+
+  return {
+    payloadErrors,
+    ruleUpgradePayloads,
+  };
+};
+
+interface GetValueForFieldArgs {
+  field: keyof PrebuiltRuleAsset;
+  upgradeableRule: UpgradeableRule;
+  globalPickVersion: PickVersionValues;
+  requestBody: PerformRuleUpgradeRequestBody;
+  ruleId: RuleSignatureId;
+  ruleFieldsDiff: AllFieldsDiff;
+}
+
+const getValueForField = ({
+  field,
+  upgradeableRule,
+  globalPickVersion,
+  requestBody,
+  ruleId,
+  ruleFieldsDiff,
+}: GetValueForFieldArgs) => {
+  const fieldUpgradeStatus = determineFieldUpgradeStatus(field, upgradeableRule);
+
+  if (fieldUpgradeStatus.type === 'NON_UPGRADEABLE_FIELD') {
+    return fieldUpgradeStatus.value;
+  }
+
+  const fieldUpgradeSpecifier = createFieldUpgradeSpecifier({
+    field,
+    requestBody,
+    globalPickVersion,
+    ruleId,
+    targetRuleType: upgradeableRule.target.type,
+    ruleFieldsDiff,
+  });
+
+  if (!fieldUpgradeSpecifier) {
+    throw new Error(`Missing field upgrade specifier for: ${ruleId}`);
+  }
+
+  if (fieldUpgradeSpecifier.pick_version === 'RESOLVED') {
+    const resolvedValue = fieldUpgradeSpecifier.resolved_value;
+
+    return getMappedValueForField(field, resolvedValue);
+  }
+
+  return getNonResolvedValueForField(field, upgradeableRule, fieldUpgradeSpecifier, ruleFieldsDiff);
+};
+
+const getNonResolvedValueForField = (
+  field: keyof PrebuiltRuleAsset,
+  upgradeableRule: UpgradeableRule,
+  fieldUpgradeSpecifier: NonNullable<RuleFieldsToUpgrade[keyof RuleFieldsToUpgrade]>,
+  ruleFieldsDiff: AllFieldsDiff
+) => {
+  const ruleId = upgradeableRule.target.rule_id;
+  const groupNameForField = getGroupNameForField(field, upgradeableRule.target.type);
+
+  if (fieldUpgradeSpecifier.pick_version === 'MERGED') {
+    const ruleDiff = ruleFieldsDiff[groupNameForField];
+
+    if (ruleDiff && ruleDiff.conflict !== 'NONE') {
+      throw new Error(
+        `Automatic merge calculation for field '${groupNameForField}' in rule of rule_id ${ruleId} resulted in a conflict. Please resolve the conflict manually or choose another value for 'pick_version'.`
+      );
+    }
+
+    const mergedVersion = ruleDiff.merged_version;
+
+    return getMappedValueForField(field, mergedVersion);
+  }
+
+  const pickVersion = fieldUpgradeSpecifier.pick_version.toLowerCase() as keyof UpgradeableRule;
+
+  // By this point, can be only 'base', 'current' or 'target
```

Comment:
At this point, the pick_version can really be only 'base', 'current' or 'target', because for the cases that `pick_version` was `RESOLVED` it was already handled with `getMappedValueForField` as you can see in `getValueForField`.

Unfortunately, TS does not realize this and does not narrow down the scope for `pick_version`. Let me know if you find a way to do this (hopefully without overcomplicating the readability of the code)